pkg/errgroup/errgroup_test.go

```diff
@@ -0,0 +1,112 @@
+// +build test_unit
+
+/*
+Copyright 2017 The Nuclio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errgroup
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/nuclio/logger"
+	nucliozap "github.com/nuclio/zap"
+	"github.com/stretchr/testify/suite"
+)
+
+type ErrGroupTestSuite struct {
+	suite.Suite
+	logger logger.Logger
+	ctx    context.Context
+	lock   sync.Mutex
+}
+
+func (suite *ErrGroupTestSuite) SetupTest() {
+	suite.logger, _ = nucliozap.NewNuclioZapTest("test")
+	suite.ctx = context.Background()
+	suite.lock = sync.Mutex{}
+}
+
+func (suite *ErrGroupTestSuite) TestSemaphoredErrGroup() {
+
+	for _, testCase := range []struct {
+		name          string
+		concurrency   int
+		goroutinesNum int
+	}{
+		{
+			name:          "DefaultConcurrency",
+			concurrency:   DefaultErrgroupConcurrency,
+			goroutinesNum: 10,
+		},
+		{
+			name:          "ZeroConcurrency",
+			concurrency:   0,
+			goroutinesNum: 10,
+		},
+		{
+			name:          "NegativeConcurrency",
+			concurrency:   -45,
+			goroutinesNum: 10,
+		},
+		{
+			name:          "ManyGoroutines",
+			concurrency:   7,
+			goroutinesNum: 30,
+		},
+	} {
+		suite.Run(testCase.name, func() {
+			var concurrentCallCount, totalCallCount int
+			errGroup, errGroupCtx := WithContext(suite.ctx, suite.logger, testCase.concurrency)
+
```

Comment:
why not doing the `increase / decrease call count` within these test cases block? more readable then condition the function base on `true/false`

Reply:
Since I'm doing both `increase` and `decrease` in every test case I figured to combine them in a single method.

I can do it explicitly for better readability.